lib/classes/PluginManager.js

```diff
@@ -638,6 +638,7 @@ class PluginManager {
 
     try {
       await this.invoke(commandsArray);
+      for (const { hook } of this.getHooks(['success'])) await hook();
```

Comment:
I'm not sure about placement of this - I wanted to makes sure that it's only ran if there are not errors, but also ensure that we run `error` hooks if `success` hook fails. Do you think we should handle it this way or not run `error` hook on `success` hook failure?

Reply:
Good point. 

I think it should be consistent with how we handle `initialize` hook, so error in it should not trigger `error` hook.

And as I inspect current logic, I find a bit surprising that we're processing backend request (showing notification) also when deploy crashes. Was that intended? Or was it accidentally added?

As look into it, I think proper logic should be as:

```javascript
await runInitializeHooks();
deferredBackendNotificationRequest = handleTelemetry()
try {
  await runNormalHooks();
} catch (error) {
  try {
    await runErrorHooks();
  } catch {
    await deferredBackendNotificationRequest;
    throw error;
  }
}

try {
  await runFinalizeHooks();
} catch (error) {
    await deferredBackendNotificationRequest;
    throw error;
}
await processBackendNotificationRequest(await deferredBackendNotificationRequest);
```

Reply:
That's good point that you bring up here with notifications - I think this might've been introduced by a mistake as I don't remember that we wanted to show notifications in case of errors - I'm also not sure how it behaved previously, do you remember?

Reply:
As I check notifications where shown only on success, although logic may not look that clear: https://github.com/serverless/serverless/blob/bd4e792594843ab51d1859f1fbc499aa4a4513c1/lib/plugins/deploy.js#L55-L63

We were reporting telemetry before _deploy_ logic (but unconditionally after _packaging_ no matter if it succeded or not), and result was passed to backend notifications processor with last hook (and that one is not triggered if there's error on the way)

Reply:
Also logic we're discussing here, takes place only if there's a local fallback from older major release.

In other cases telemetry is handled in main script, and over there notification is only shown on success.
So we can treat it as bug here, that we're doing otherwise

Reply:
Thanks for the clarification, that sounds like a bug to me then - should I adjust it here to only show notification in situations where the command did not throw an error?

Reply:
Yes I would definitely do it. It can be two commits (one that fixes issue, and second that adds handling of final hook)